contrib/mpi-proxy-split/split_process.cpp

```diff
@@ -328,6 +332,67 @@ startProxy()
   return childpid;
 }
 
+// Rounds the given address up to the nearest 2MB.
+static inline unsigned long long
+alignMemAddr(unsigned long long addr)
+{
+  return (addr + 0x1fffff) & ~0x1fffff;
+}
+
+// This sets the address range of the lower half by searching through the memory
+// region for the first available free region.
+static void
+dynamicMemRangeHelper(MemRange_t *lh_mem_range, const uint64_t *region_size)
+{
+  const uint64_t ONE_GB = 0x40000000;
+  
+  bool is_set = false;
+
+  char perms[8];
+  unsigned long offset;
+  char device[8];
+  unsigned long inode;
+  char prev_path_name[PATH_MAX];
+  char next_path_name[PATH_MAX];
+
+  unsigned long long prev_addr_start;
+  unsigned long long prev_addr_end;
+  unsigned long long next_addr_start;
+  unsigned long long next_addr_end;
+
+  // FIXME: fopen and fscanf may call malloc, which might call mmap.
+  // For now, we hope it doesn't change the calculation.
+  FILE* f = fopen("/proc/self/maps", "r");
+  if (f) {
+    fscanf(f, "%llx-%llx %s %lx %s %lx %s\n",
+           &prev_addr_start, &prev_addr_end, perms, &offset, device, &inode, prev_path_name);
+    while (fscanf(f, "%llx-%llx %s %lx %s %lx %s\n",
+           &next_addr_start, &next_addr_end, perms, &offset, device, &inode, next_path_name) != EOF) {
+      // alignMemAddr aligns the address such that HUGEPAGES/2MB can also be used.
+      prev_addr_end = alignMemAddr(prev_addr_end);
+
+      // We add a 1GB buffer between the end of the heap or the start of the stack.
+      if (strcmp(prev_path_name, "[heap]") == 0) {
+        prev_addr_end += ONE_GB;
```

Comment:
It's confusing to check both `prev` and `next` in this while loop.
I think we only need to check `current`, which is the line just read from `proc/maps`
Whatever you need to do for `current`, will be copied to `prev`.

Reply:
right, i think that would only be a problem if the very first line read is the heap memory region - is that possible?